[Blog] April community update (#199)
pulsar-edit/pulsar-edit.github.io: docs/blog/20230401-Daeraxa-AprUpdate.md

---
title: Community Update
author: Daeraxa
date: 2023-04-01
category:
  - news
  - log
tag:
  - update
---

Another dose of our regular monthly community update!

<!-- more -->

# Welcome to the April Community Update (don't worry, no fooling going on here)!

> Hi everyone and welcome to the April installment of our monthly update that keeps keeps you informed about what is going on in the background of Pulsar, work in progress, community contributions and more!

Reviewer: ```suggestion
Hi everyone and welcome to the April installment of our monthly update that keeps you informed about what is going on in the background of Pulsar, work in progress, community contributions and more!
```

---
title: Community Update
author: Daeraxa
date: 2023-04-01
category:
  - news
  - log
tag:
  - update
---

Another dose of our regular monthly community update!

<!-- more -->

# Welcome to the April Community Update (don't worry, no fooling going on here)!

> Hi everyone and welcome to the April instalment of our monthly update that keeps keeps you informed about what is going on in the background of Pulsar, work in progress, community contributions and more!

Reviewer 2: ```suggestion
Hi everyone and welcome to the April installment of our monthly update that keeps keeps you informed about what is going on in the background of Pulsar, work in progress, community contributions and more!
```

Reviewer 3: This one's a UK/US difference, so maybe we flip a coin? ;-)

Author: Yeah but I guess its meant to be from "the team" rather than me specifically so it reverts to en-US rules, didn't even know that was a difference...

> Hi everyone and welcome to the April instalment of our monthly update that keeps keeps you informed about what is going on in the background of Pulsar, work in progress, community contributions and more!

Reviewer: ```suggestion
Hi everyone and welcome to the April installment of our monthly update that keeps keeps you informed about what is going on in the background of Pulsar, work in progress, community contributions and more!
```

Author: @Spiker985 mentioned this one above which has been committed. (Basically it seems its UK vs US spelling but as its meant to be from the team then it goes to US spelling, didn't know there was even a difference...)

Author: I've made some changes to this section as well to reflect that we already added a "good" badge along with an example.

Reviewer: I appreciate all of that, I've applied my final review here, so I'll close this one out

You will notice I've tried to jazz the post up a little with some images this time around, please [let us know](https://pulsar-edit.dev/community.html) if you have any comments on this addition.

Now that has been dealt with, onto the updates!

## i18n (internationalization) Efforts

![303535.png](:/a303a122a01f4662bbb60789862f1eda)[^1]
[@meadowsys] and [@confused-techie] have been making great progress on the efforts to provide native i18n functionality within Pulsar. This means you will be able to select your own locale/language and Pulsar will be able to translate the various menus and items for you. Of course the big caveat here is that currently none of it is translated so if you have the ability to then we will be asking the community to provide translations wherever possible. To do this we are currently planning to use [Crowdin](https://crowdin.pulsar-edit.dev/). We will make a bigger announcement on this feature once it is ready.

## Tree-sitter Modernization

![747df1a39d46487ecc42af0dd66d59bc.png](:/f75d48ef89a84a02ad6bfe3a203fa39d)[^2]

> Another month and more progress has been made on this front, particularly by [@maurício szabo] and [@savetheclocktower], to get our tree-sitter implementation modernized. Some users noticed that opening some languages on Pulsar, like C++, Java, and Ruby, makes some syntax highlighting to get different from what they were in Atom, and most of the time, highlighting is either wrong or some tokens are simply missing.

Reviewer 3: ```suggestion
Another month and more progress has been made on this front, particularly by [@maurício szabo] and [@savetheclocktower], to get our tree-sitter implementation modernized. Some users noticed that opening some languages on Pulsar, like C++, Java, and Ruby, makes some syntax highlighting tokens different from what they were in Atom, and most of the time, highlighting is either wrong or some tokens are simply missing.
```

> With this rewrite, we basically are moving syntax highlight to the modern version of tree-sitter, using queries. This means some tokens will be more precise, we'll not have a "different implementation" of all other editors, and we can keep up-to-date with the recent developments and advancements on tree-sitter

Reviewer 3: ```suggestion
With this rewrite, we're updating Atom’s tree-sitter support to use a newer approach for mapping tree nodes to scope names: using [queries](https://tree-sitter.github.io/tree-sitter/using-parsers#pattern-matching-with-queries). This means more accurate mapping of scopes. We can also use queries to define indentation rules, code folding, and other features. It further means we won’t be stuck with a "different" implementation of tree-sitter from all other editors, and we can keep up-to-date with the recent developments and advancements on tree-sitter.
```

Author: Sounds great!

For more background info see some of the previous updates or [Maurício's blog post](https://pulsar-edit.dev/blog/20230209-mauricioszabo-tree-sitter-part-1.html) on the topic.

## TextMate Grammar Library & Superstring Migration to WASM

![240px-WebAssembly_Logo.svg.png](:/56a284d5f7b8418c958f0be0332831b5)[^3]
Atom and current versions of Pulsar use a library called [first-mate](https://github.com/pulsar-edit/first-mate) that uses [node-oniguruma](https://github.com/pulsar-edit/node-oniguruma/) to parse the legacy TextMate grammars (the original ones before Atom moved to tree-sitter as its primary grammar choice). Like tree-sitter and superstring, this is something that is preventing our migration to modern versions of Electron so [@maurício szabo] and [@savetheclocktower] have been working on this to instead migrate to a new library we are calling `second-mate` which uses [vscode-oniguruma](https://github.com/microsoft/vscode-oniguruma) instead which is WASM based. Funny how things eventually come back around to borrowing from VSC instead...

> Work has also been going on to migrate the existing [superstring](https://github.com/pulsar-edit/superstring) library to WASM. This, along with `tree-sitter` and `first-mate` mentioned above is blocking our progress to modern Electron versions.

Reviewer 3: ```suggestion
Work has also been going on to migrate the existing [superstring](https://github.com/pulsar-edit/superstring) library to WASM. This, along with `tree-sitter` and `first-mate` mentioned above, is blocking our progress to modern Electron versions.
```

## Package Service Filtering

![216314.png](:/8854b3fce3f640f88b90ede400574b63)[^4]
An interesting update to our backend that allows for new queries to be made for finding out what packages provide or consume a given service. This can be used in a number of ways, users can easily filter by package functionality rather than just keywords and developers/packages authors can search to see what packages may provide or consume a given service to help create a more interconnected set of packages.

For example this query [https://web.pulsar-edit.dev/packages?service=terminal&serviceType=provided](https://web.pulsar-edit.dev/packages?service=terminal&serviceType=provided) will search for all packages that provide the `terminal` service so if somebody wishes to find packages that provide or consume `terminal` then it is now easy to do so.

This functionality is currently only available via the API and website query but we plan to offer some interesting and (hopefully) useful features on the website (and potentially Pulsar itself) to display these services.

## Bundling `ppm` Within Pulsar

![node-js-icon-20.png](:/6fd566d5eec9436b9425cfae3038d335)[^5]

> As you may know, `ppm` is the `Pulsar Package Manager`, i.e. the thing you interact with when you want to install or update packages. Most people will probably only use this by interacting with the packages menu inside Pulsar but `ppm` is also supplied as a command line applications accessible via either `ppm` or `pulsar -p`. One issue with this approach is that we have to account for all of the different OSs and packaging methods which means adding it to `$PATH` correctly on each OS etc. (This was even an issue with Atom which had a special macOS only command to install the `apm` command line again).

Reviewer 3: ```suggestion
As you may know, `ppm` is the `Pulsar Package Manager` —  the thing you interact with when you want to install or update packages. Most people will probably only use this by interacting with the packages menu inside Pulsar but `ppm` is also supplied as a command line applications accessible via either `ppm` or `pulsar -p`. One issue with this approach is that we have to account for all of the different OSs and packaging methods which means adding it to `$PATH` correctly on each OS etc. (This was even an issue with Atom which had a special macOS only command to install the `apm` command line again).
```
